src/runtime_src/core/common/usage_metrics.h

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+#ifndef XRT_CORE_USAGE_METRICS_H
+#define XRT_CORE_USAGE_METRICS_H
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <stdint.h>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "core/include/ert.h"
+#include "core/include/xrt.h"
+#include "core/include/xrt/xrt_hw_context.h"
+#include "core/include/xrt/xrt_uuid.h"
+
+// forward declaration of xrt_core::device class
+namespace xrt_core {
+class device;
+}
+
+namespace {
+using tp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+}
+
+////////////////////////////////////////////////////////////////
+// namespace xrt_core::usage_metrics
+//
+// Used for printing XRT usage metrics of Application.
+//
+// This information is printed at the end of the application run
+// and the infrastructure must be initialized before launching the
+// application using xrt.ini
+//
+// % cat xrt.ini
+// [Runtime]
+// usage_metrics_logging = true
+////////////////////////////////////////////////////////////////
+namespace xrt_core::usage_metrics {
+
+struct bo_metrics
+{
+  uint32_t total_count = 0;
+  size_t   total_size_in_bytes = 0;
+  size_t   peak_size_in_bytes = 0;
+  size_t   bytes_synced_to_device = 0;
+  size_t   bytes_synced_from_device = 0;
+};
+
+struct kernel_metrics
+{
+  struct timestamp
+  {
+    tp start_time;
+    bool is_valid = false;
+  };
+
+  std::string name;
+  std::vector<uint32_t> cu_index_vec;
+  uint32_t total_runs = 0;
+  std::chrono::microseconds total_time = {};
+  std::unordered_map<uintptr_t, timestamp> exec_times; // run handle ptr is used for indexing
+  size_t num_args;
+};
+
+struct hw_ctx_metrics
+{
+  uintptr_t id;  // using hw_ctx handle ptr as unique identifier for logging
+  xrt::uuid xclbin_uuid;
+  struct bo_metrics bos_met;
+  std::vector<struct kernel_metrics> kernel_metrics_vec;
+};
+
+struct device_metrics
+{
+  std::string bdf = "";
+  struct bo_metrics global_bos_met;
+  uint32_t bo_active_count = 0;
+  uint32_t bo_peak_count = 0;
+  std::vector<struct hw_ctx_metrics> hw_ctx_vec;
+};
+
+using metrics_map = std::map<std::thread::id, std::map<unsigned int, device_metrics>>;
```

Comment:
I don't think any of these structs are needed in the header file?
Can they be moved to the .cpp file?
This goes for the concrete `class usage_metrics_logge` also, it can be moved to .cpp.
We really don't want to expose the implementation of the metrics to  compilation units that include this header.